text/0000-target-tier-policy.md

```diff
@@ -0,0 +1,222 @@
+- Feature Name: `target_tier_policy`
+- Start Date: 2019-09-20
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+We should have an official, objective policy for adding new (tier 3) targets,
+and for raising targets to tier 2 (with rustup builds) or even tier 1.
+
+# Motivation
+[motivation]: #motivation
+
+Rust developers regularly implement new targets in the Rust compiler, and
+reviewers of pull requests for such new targets would like a clear, consistent
+policy to cite for accepting or rejecting such targets. Currently, individual
+reviewers do not know what overall policy to apply, and whether to apply solely
+their own judgment or defer to a Rust governance team.
+
+Rust developers regularly ask how they can raise an existing target to tier 2
+(and in particular how they can make it available via rustup), and occasionally
+ask what it would take to add a new tier 1 target. The Rust project has no
+clear policy for target tiers. People not only don't know, they don't know who
+to ask or where to start.
+
+(See https://forge.rust-lang.org/platform-support.html for more information
+about targets and tiers.)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+At a high level, the three tiers break down as follows:
+
+- Tier 3 targets provide no guarantees of support.
+- Tier 2 targets will always build, but may not pass tests.
+- Tier 1 targets will always build and pass tests.
+
+Adding a new tier 3 target imposes minimal requirements; we focus primarily on
+avoiding disruption to other ongoing Rust development.
+
+Tier 2 and tier 1 targets place work on the Rust community as a whole, to avoid
+breaking the target. Thus, these tiers require commensurate efforts from the
+maintainers of the target, to demonstrate value and to minimize any disruptions
+to ongoing Rust development.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
```

Comment:
I think this is a good point to make it clear that you should meet the requirements for a tier before opening the asoociated promotion PR.

Something like:

```
You should normally ensure that your target meet the tier requirements before posting a PR. In general, we would prefer to add a target to a tier once it is clear that the target has been meeting the standards set for the tier for a significant time period.
```

Reply:
> for a significant time period.

If we add this language, we should define in this RFC what the time period is. A release cycle? 2 release cycles?

Reply:
That time should ideally also be a function of the tier, with more time for tier 1.

Reply:
I would expect something more, like 6 months, which means yes we need to be more specific it two people are picking very different assumptions.

Reply:
What would a PR promoting a tier look like? As far as I know there's no tight coupling between a tier and the code and infrastructure. It's not like have a TOML file that enforces the tier structure, it's mainly a social commitment. You could make a PR updating the documentation on forge, but this doesn't seem like best place to have the promotion discussion.

It would also be reasonable for the procedure for promoting or demoting to be out of scope of this RFC. Since we've never done this before I think it might be best to create the procedure when we actually want to promote a platform, rather than discuss and try to reason around a hypothetical promotion.

Reply:
> What would a PR promoting a tier look like? As far as I know there's no tight coupling between a tier and the code & infrastructure. It's not like we have a TOML file that enforces the tier structure, it's mainly a social commitment. You could make a PR updating the documentation on forge, but this doesn't seem like best place to have the promotion discussion.
> 
> It would also be reasonable for the procedure for promoting or demoting to be out of scope of this RFC. Since we've never done this before I think it might be best to create the procedure when we actually want to promote a platform, rather than discuss and try to reason around a hypothetical promotion.

The bit of my comment I care about is telling people that promotion to a higher tier (or addition to tier 3) trails meeting the requirements.

In other words, let's call out that you not only can meet the standards for any of the support tiers without being promoted into that tier, but that you should expect to meet the standards first, for long enough that people watching the process can look at your historic behaviour, and then go into the tier promotion process, whatever that might be.

Reply:
In the past I've been told that the bar for Tier 3 is as low as "file a PR to get your target specification into rustc's files, and make sure `core` can build".

Reply:
> What would a PR promoting a tier look like?

Isn't that obvious? it's a PR that changes the CI such that it now gates changes to master in accordance with the tier policy. Like https://github.com/rust-lang/rust/pull/57130.

Reply:
@jethrogb Your answer doesn't actually describe what that PR would be though. What does a person need to use, where do they need to add it, how can they test it, where do they document it, who do they ask to review it?

Your PR example for instance is actually a good example of it's not so obvious.

As I understand that PR only makes fortanix a tier 2 **target** platform, if I wanted to make it so I could build rustc, cargo, or tools on fortanix, what would I need to do?

Elsewhere in this thread, team members have said that promotion should go through the infra, release, and compiler teams. However that wasn't the process that, that PR went through, so who should I be asking for a review for promotion?

I think the answers to these questions are only obvious if you have picked up on this information through osmosis, which is why it should be documented. Though as I said I don't this needs to be included with this policy though, it could be added later.

Reply:
As I mentioned a while ago elsewhere in the comments, some targets will never be **hosts** for rustc/cargo but should be totally fine to **cross-compile** to. This is practically the default situation for embedded targets. So we absolutely need to get this distinction codified somewhere.

Reply:
Yeah that's fair, there are definitely `std`-only (and presumably `core`-only or `core+alloc`-only) tier 2 platforms (and perhaps one day tier 1 ?)  But once there's `rustc` and `cargo` people start expecting `rustup` too.

Reply:
Well, long term, `rustup` should (ideally) allow you to install the necessary target files so that you can cross-build to the target without needing to go all the way to `cargo xbuild`.

Maybe I'm not understanding the terms you're using, but I'd say that:
* `rustup` providing a copy of the `core`  crate is the minimum possible official support for a platform.
* Then it might also provide `alloc` on top of that
* Then it might also provide `std` on top of that.
* Then it might also provide installs of `rustc` and `cargo` and any other rust tools (rustfmt, clippy, whatever) that run within the target environment.

But each item in the list fairly strictly requires all previous items in the list.

Reply:
I think you're describing `rustup target add footarget` which already works for the core/alloc/std stuff.  `rustc` and `cargo` imply that it's possible to *host* on that platform at which point people want to be able to have `rustup` on their platform and expect `rustup toolchain install stable` to DTRT.

Reply:
ohhhhhhhhhhhhh

yes, okay. Then my concern isn't related to what you're talking about. Having rustup hosted on the target is totally separate from having rustup hosted elsewhere cross-built to the target. Nevermind me :3

Reply:
I've added language requiring that a target meet the requirements before posting an issue or PR to propose the target for introduction or promotion.

I've also added an entry in the "alternatives" section noting that we *could* require a concrete amount of time to pass. I don't think we should have such a requirement, but I do think it's worth capturing as an alternative.